### DISCUSSION TITLE
Recoil state through Redux devtool with Replay and Time travel

### DISCUSSION THREAD

**Discussion by theohagos** [facebookexperimental/Recoil#1649] [Show and tell]
Here is the package, give it a try : https://www.npmjs.com/package/@theohagos/recoil-observer

**Comment by snayerman** [comment]
This is awesome, thanks for this! I made a very simple debugger a while back to log changes in particular atoms, but real devtools like these are what is sorely missing from recoil.

**Reply by theohagos** [reply]
Thank you, appreciate the comment. It was possible because both Recoil and Redux-devtool have wonderful APIs to allow us to do this.

**Comment by SmashPlusUltra** [comment]
If you have already redux in your project, would you'll be able to use this package anyways? Or I can't be use in conjunction with redux?

**Reply by theohagos** [reply]
Yes, but will only show recoil states.

**Comment by theohagos** [comment]
Sure you can but this will only show the recoil states.

On Fri, 8 Apr 2022 at 06:34, SmashPlusUltra ***@***.***>
wrote:

> If you have already redux in your project, would you'll be able to use
> this library anyways? Or I can't be use in conjunction with redux?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/facebookexperimental/Recoil/discussions/1649#discussioncomment-2528510>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACAQTOXBHDBFPTCIJGKNQUDVD7AN3ANCNFSM5P2J2YMA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> .com>
>


-- 
Regards,
Theo

**Reply by SmashPlusUltra** [reply]
I've installed but I can't get any recoil state just redux states, what could I check to see what happens?
My app.js looks like
```
import React, { lazy, Suspense, useContext } from 'react';
import './App.scss';
import { BrowserRouter as Router} from 'react-router-dom';
import Login from './Layout/Login';
import NotificationPrincipal from './Messages/NotificationPrincipal'
import ChatBotHOC from './utils/ChatBotHOC';
import { Spinner } from 'react-bootstrap';

import ErrorBoundary from './utils/ErrorBoundary'
import { RecoilRoot } from 'recoil'
import SignalR from './Hooks/SignalRConnector'
import { Provider as ApiProvider } from './Context/apiContext';
import { RecoilObserver } from '@theohagos/recoil-observer';

export default function App() {
    return (
        <ErrorBoundary>
            <Suspense fallback={<Spinner variant="success" animation="grow" />}>                
                <RecoilRoot>
                    <RecoilObserver/>
                    <ApiProvider>
                        <SignalR>
                            <Router>
                                <div className="App">
                                    <NotificationPrincipal />
                                    <Login />
                                    <ChatBotHOC />  
                                </div>                                                           
                            </Router>
                        </SignalR>
                    </ApiProvider>
                    </RecoilRoot>                    
            </Suspense>
        </ErrorBoundary>
    );
}
```

Thanks!

**Reply by theohagos** [reply]
are you able to see "Recoil State Observer" on the Redux instance selector drop down? That needs to be selected in order to see recoil states on Redux.

**Reply by SmashPlusUltra** [reply]
Hi!, no... just redux.

**Reply by theohagos** [reply]
can you post a screenshot of the redux devtool dropdown menu expanded, please?

**Reply by SmashPlusUltra** [reply]
Sure!  All that you can see comes from redux.
![image](https://user-images.githubusercontent.com/53838441/164245531-4f3211b1-eb03-4115-9268-2b25c77a704f.png)

In the index.js I have the following
```
ReactDOM.render(        
    <React.StrictMode>
        <Provider store= {store}>//Redux store
            <App />
        </Provider>
  </React.StrictMode>,
        document.getElementById('root')
    );
```
Could it be that because the redux store is configure before the observer the first one is getting the output in the redux  devtool dropdown? I'll try to prove it.

**Reply by theohagos** [reply]
Yes, that is a very good observation,  where I work we have an app that has the same setup as you, and I just checked and <RecoilRoot> wraps Redux provider and we are able to see both recoil and redux states.

**Comment by SmashPlusUltra** [comment]
Could it be possible that the oberver produce some memory leak? I have a heavily intense websocket client that populate a table and make updates in real-time and it goes from 700..999mb consume to 3..4gb or even more.
I'm not sure the observer was the problem, just I remove it after doing some memory analysis with devtools, and it  works ok after that... and after I put it back to check and the problem arise again.
Maybe if you don't have hundreds of messages coming to your app trying to update de UI you don't notice it.

**Reply by theohagos** [reply]
Hello @SmashPlusUltra that is certainly a possibility but the Observer doesn't keep any state, it is just a listener and publisher of events between recoil and redux dev tool. I suspect the developer tool is consuming a lot of memory. You can try and restrict which states are shown by passing a list of the keys to the observer and see if it helps at all?

**Reply by theohagos** [reply]
Oh forgot to ask that you did manage to get it to show the states after all, I know you had some issues previously?

**Comment by drarmstr** [comment]
@theohagos If this solution is working well for people, would it make sense to reference it in the [documentation](https://recoiljs.org/docs/guides/dev-tools)?  (cc @mondaychen @butlersrepos)

**Reply by mondaychen** [reply]
Sure! It's certainly a good thing to let users be aware of handy tools

**Reply by theohagos** [reply]
Hello, @drarmstr and @mondaychen More than happy to do that, but how do we get it in there?

By the way, we are using recoil in our production apps and I put this together to help our developers view application states and is working really well so decided to publish it on npm.

**Reply by mondaychen** [reply]
Can you send a PR to the docs branch?

**Reply by theohagos** [reply]
OK, will do thanks
